src/core/public/keyboard_shortcut/key_parser.ts

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { VALID_KEY_STRING_REGEX } from './utils';
+import { ALLOWED_KEYS, CODE_TO_KEY_MAPPING, DISPLAY_MAPPINGS } from './constants';
+
+export class KeyStringParser {
+  /**
+   * Normalizes a key string from shortcut definition into consistent lowercase format
+   *
+   * Validates and converts key strings to lowercase for consistent registration.
+   * All platforms register shortcuts the same way but display them appropriately
+   * for each platform.
+   *
+   * @param keyString - Raw key combination string (e.g., "cmd+s", "Alt+Shift+A")
+   * @returns Normalized lowercase key string
+   * @throws Error when input is invalid or malformed
+   *
+   * @example
+   *
+   * // Valid key combinations are normalized to lowercase:
+   * parser.normalizeKeyString("CMD+S") // → "cmd+s"
+   * parser.normalizeKeyString("Alt+H") // → "alt+h"
+   * parser.normalizeKeyString("cmd+shift+a") // → "cmd+shift+a"
+   *
+   * // Invalid keys are rejected:
+   * parser.normalizeKeyString("f1") // → throws Error
+   * parser.normalizeKeyString("ctrl+s") // → throws Error (use "cmd+s")
+   * parser.normalizeKeyString("cmd++") // → throws Error
+   *
+   */
+  public normalizeKeyString(keyString: string): string {
+    const lowercasedKeyString = keyString.toLowerCase();
+    if (!VALID_KEY_STRING_REGEX.test(lowercasedKeyString)) {
+      throw new Error(
+        `Invalid key combination: "${keyString}". ` +
+          `Use supported keys (a-z, 0-9, punctuation, arrows, special keys) with valid modifiers (cmd, alt, shift). ` +
+          `Examples: "cmd+s", "alt+shift+a", "g+d"`
```

Comment:
g+d should not be allowed. And what if they wrote `shift+alt+a` and is getting this error? I would just be vague:

`Invalid key combination: ${keyString}. Please refer to our documentation to see the allowed formats`